Make Inria's CI main job blue again

shindere opened the pull request:
This PR achieves this by skipping those configurations on which OCaml is known not to work.

It's not a matter of hiding the problmes. The benefits of this PR are:

 1. Enabling this job can help detecting regresions in the backends.
 2. It provides a list of ploatorms not supported.
 3. The list is version-controlled so it can be updated in synch wiht the changes that restore the suport on a given platform.

No attempt to make the code short or concise or clever has been made.
The PR does rather favour readability and maintainability.

shindere commented:
By the way, see [precheck build #649](https://ci.inria.fr/ocaml/job/precheck/build/649)

gasche commented on testsuite/tests/backtrace/backtrace_effects_nested.flambda.reference:
```diff
@@ -1,4 +1,4 @@
 Raised by primitive operation at Backtrace_effects_nested.blorp in file "backtrace_effects_nested.ml", line 21, characters 2-11
-Called from Stdlib__EffectHandlers.Deep.continue in file "effectHandlers.ml" (inlined), line 36, characters 21-62
+Called from Stdlib__Effect.Deep.continue in file "effect.ml" (inlined), line 36, characters 21-62
```
This change was already part of #10912 which I just mrged.

shindere commented:
Rebased.

dra27 commented on tools/ci/inria/main:
```diff
@@ -45,6 +45,44 @@ arch_error() {
+  ocaml-linux-32) unsupported_config;;
+  ocaml-linux-64) unsupported_config;;
+  ocaml-mingw-32) unsupported_config;;
+  ocaml-mingw-64) unsupported_config;;
```
ocaml-mingw-64 is a supported config!

dra27 commented on tools/ci/inria/main:
```diff
@@ -45,6 +45,44 @@ arch_error() {
+unsupported_config()
+{
+  if ${flambda}; then
+    withflambda=" with FLambda ";
```
Nit (typo): stray capital "L" in Flambda

dra27 replied:
Nit (display): there should be no space before with (given its use below)

dra27 reviewed (commented):
mingw-w64 is a working configuration, modulo one test which has open PRs.

It's clearly better not to be generating the noise of failure logs on each build, but I'm not personally that thrilled by fake passes in CI (strangely, @shindere, Jenkins displays passes in blue!!).

Can we not alter the matrix configuration to skip these jobs as some existing jobs already are - I had the briefest of looks in the configuration last week and the only question that needed resolving was if one could refer to the branch name in the skip expression which is already to skip, say, flambda on arm32 (and various Windows workers) so that you could have `branch == "trunk" && (node-name = etc.)`

dra27 commented:
It occurs to me that suggesting that the Jenkins configuration be altered of course hides a benefit of this PR of having a list of the known failing platforms at the moment. However, we have a list of supported platforms in the README which should regardless be updated with the multicore status.

xavierleroy commented:
> Can we not alter the matrix configuration to skip these jobs as some existing jobs already are - I had the briefest of looks in the configuration last week and the only question that needed resolving was if one could refer to the branch name in the skip expression which is already to skip, say, flambda on arm32 (and various Windows workers) so that you could have branch == "trunk" && (node-name = etc.)

An issue with this approach is that the list of OCaml 5.00 supported configurations is in 3 different places:
- the configure script (versioned)
- the Jenkins config for job "main" (unversioned, not accessible)
- the Jenkins config for job "precheck" (unversioned, not accessible).

To me it's a maintenance nightmare.

@shindere's proposal has the list in "only" two places:
- the configure script (versioned)
- tools/ci/inria/main script

It's better but still redundant.

My proposal is to have the info exclusively in the configure script:  the CI main script calls configure and if the latter exits with  a "configuration not supported" error code, the test is skipped.

shindere commented:
Many thanks for your participation to the discussion Xavier, well
appreciated.

Xavier Leroy (2022/01/19 02:28 -0800):
> > Can we not alter the matrix configuration to skip these jobs as some existing jobs already are - I had the briefest of looks in the configuration last week and the only question that needed resolving was if one could refer to the branch name in the skip expression which is already to skip, say, flambda on arm32 (and various Windows workers) so that you could have branch == "trunk" && (node-name = etc.)
>
> An issue with this approach is that the list of OCaml 5.00 supported configurations is in 3 different places:
> - the configure script (versioned)
> - the Jenkins config for job "main" (unversioned, not accessible)
> - the Jenkins config for job "precheck" (unversioned, not accessible).

I think tha's not completely true because what precheck does is that it
actually runs the main job. It's just that it's parameterised by the
repo and branch on which to run it but, as far as I can remember, apart
from that we are really running the main job itself.

> To me it's a maintenance nightmare.

I agree. And I confirm the accessibility problems with maintaining the
matrix in Jenkins, that's something I won't be able to deal with on my
own.

> @shindere's proposal has the list in "only" two places:
> - the configure script (versioned)
> - tools/ci/inria/main script

Whcih is versionned, too.

> It's better but still redundant.
>
> My proposal is to have the info exclusively in the configure script:
> the CI main script calls configure and if the latter exits with  a
> "configuration not supported" error code, the test is skipped.

But wiht your proposal, it makes it harder to work on restoring the
suuport for those configurations where it has been broken. With the
approach here it remains possible and easy to configure for the
unsupported platforms and work on them.

To be more precise: yes, I think it can be useful, for releases, tha
tconfigure fails on platforms that are not supported. But I find this
feature not so useful and perhaps even a bit counterproductive on trunk.
So that's something I would add to configure on release branches.

I'd really like to have this approved, actually.

shindere commented:
David Allsopp (2022/01/18 01:21 -0800):
> @dra27 commented on this pull request.
>
> mingw-w64 is a working configuration, modulo one test which has open
> PRs.

Can you please link them here?

> It's clearly better not to be generating the noise of failure logs on
> each build, but I'm not personally that thrilled by fake passes in CI
> (strangely, @shindere, Jenkins displays passes in blue!!).

I fixed the title of the PR accordingly, thanks. :)

> Can we not alter the matrix configuration to skip these jobs as some
> existing jobs already are - I had the briefest of looks in the
> configuration last week and the only question that needed resolving
> was if one could refer to the branch name in the skip expression which
> is already to skip, say, flambda on arm32 (and various Windows
> workers) so that you could have `branch == "trunk" && (node-name =
> etc.)`

Well that expression is rather complex already and difficult to modify
for me.

Having the control over what is tested in an accessible and versionned
way is quite useful, in my opinion.

Ultimately the main job should be rewritten to use pipelines but it's
not that easy to do for matrix jobs (my Groovy skills are not that great
at the moment)

Also, keep in mind that such an approach gives you more granularity
because it can be branch-specific, which is not that easily achievable
when the configuration is done in Jenkins' UI.

> ocaml-mingw-64 is a supported config!

It was not when I tried. If it now is I'll definitely ammend the list of
unsupported platforms.


>
> > @@ -45,6 +45,44 @@ arch_error() {
>    error "$msg"
>  }
>
> +unsupported_config()
> +{
> +  if ${flambda}; then
> +    withflambda=" with FLambda ";
>
> Nit (typo): stray capital "L" in Flambda

I removed the whole block, I intended to do so but forgot, sorry. The
branch has been rebased, too.


>
> > @@ -45,6 +45,44 @@ arch_error() {
>    error "$msg"
>  }
>
> +unsupported_config()
> +{
> +  if ${flambda}; then
> +    withflambda=" with FLambda ";
>
> Nit (display): there should be no space before with (given its use below)
>

Whole block removed.

shindere commented:
David Allsopp (2022/01/18 09:24 +0000):
> It occurs to me that suggesting that the Jenkins configuration be
> altered of course hides a benefit of this PR of having a list of the
> known failing platforms at the moment.

And a one which is branch-specific.

dra27 commented:
https://github.com/ocaml/ocaml/pull/10903 has been merged, which means that mingw-w64 is supposed to pass the testsuite

shindere committed:
Disable Inria's CI on currently unsupported platforms

shindere commented:
David Allsopp (2022/01/20 12:39 -0800):
> https://github.com/ocaml/ocaml/pull/10903 has been merged, which means
> that mingw-w64 is supposed to pass the testsuite

Thanks alot for letting me know @dra27. I have rebased and updated the
current PR to take this into account.

damiendoligez commented:
@shindere 
> But wiht your proposal, it makes it harder to work on restoring the suuport for those configurations where it has been broken. With the approach here it remains possible and easy to configure for the unsupported platforms and work on them. To be more precise: yes, I think it can be useful, for releases, tha tconfigure fails on platforms that are not supported. But I find this feature not so useful and perhaps even a bit counterproductive on trunk. So that's something I would add to configure on release branches. I'd really like to have this approved, actually.

I'm not sure I understand. This is the workflow I have in mind for restoring a new platform:
1. Check out the repo
2. patch configure to allow the new platform
3. make it work
4. patch the CI script to allow the new platform
5. launch a precheck

As far as I can tell, with @xavierleroy's proposal, step 4 simply disappears.

damiendoligez commented:
@dra27 
> It occurs to me that suggesting that the Jenkins configuration be altered of course hides a benefit of this PR of having a list of the known failing platforms at the moment. However, we have a list of supported platforms in the README which should regardless be updated with the multicore status.

Maintaining the list in the Jenkins configuration means that only people with write rights to Jenkins can do a precheck of a work-in-progress restoring a given platform. *And* it will interfere with everyone's work.

shindere commented:
Damien Doligez (2022/01/21 06:18 -0800):
> @shindere
> > But wiht your proposal, it makes it harder to work on restoring the suuport for those configurations where it has been broken. With the approach here it remains possible and easy to configure for the unsupported platforms and work on them. To be more precise: yes, I think it can be useful, for releases, tha tconfigure fails on platforms that are not supported. But I find this feature not so useful and perhaps even a bit counterproductive on trunk. So that's something I would add to configure on release branches. I'd really like to have this approved, actually.
>
> I'm not sure I understand. This is the workflow I have in mind for restoring a new platform:
> 1. Check out the repo
> 2. patch configure to allow the new platform
> 3. make it work
> 4. patch the CI script to allow the new platform
> 5. launch a precheck
>
> As far as I can tell, with @xavierleroy's proposal, step 4 simply
> disappears.

Yes, I think you are correct. But with this PR it's step 2 which
disappears.

dra27 commented:
The main job has been blue (with a tinge of yellow for MSVC and the ancient CentOS) for a little while 🙂
